openfl/transport/grpc/director_client.py

```diff
@@ -0,0 +1,281 @@
+
+"""Director clients module."""
+
+import logging
+import os
+import shutil
+import time
+from datetime import datetime
+from subprocess import check_call
+from sys import executable
+
+import grpc
+
+from openfl.pipelines import NoCompressionPipeline
+from openfl.protocols import director_pb2
+from openfl.protocols import director_pb2_grpc
+from openfl.protocols.utils import construct_model_proto
+from openfl.protocols.utils import deconstruct_model_proto
+
+
+logger = logging.getLogger(__name__)
+
+
+class ShardDirectorClient:
+    """The internal director client class."""
+
+    def __init__(self, director_uri, shard_name, disable_tls=False,
+                 root_ca=None, key=None, cert=None) -> None:
+        """Initialize a shard director client object."""
+        self.shard_name = shard_name
+        options = [('grpc.max_message_length', 100 * 1024 * 1024)]
+        if disable_tls:
+            channel = grpc.insecure_channel(director_uri, options=options)
+        else:
+            if not (root_ca and key and cert):
+                raise Exception('No certificates provided')
+            try:
+                with open(root_ca, 'rb') as f:
+                    root_ca_b = f.read()
+                with open(key, 'rb') as f:
+                    key_b = f.read()
+                with open(cert, 'rb') as f:
+                    cert_b = f.read()
+            except FileNotFoundError as exc:
+                raise Exception(f'Provided certificate file is not exist: {exc.filename}')
+
+            credentials = grpc.ssl_channel_credentials(
+                root_certificates=root_ca_b,
+                private_key=key_b,
+                certificate_chain=cert_b
+            )
+            channel = grpc.secure_channel(director_uri, credentials, options=options)
+        self.stub = director_pb2_grpc.FederationDirectorStub(channel)
+
+    def report_shard_info(self, shard_descriptor) -> bool:
+        """Report shard info to the director."""
+        logger.info('Send report AcknowledgeShard')
+        # True considered as successful registration
+        shard_info = director_pb2.ShardInfo(
+            shard_description=shard_descriptor.dataset_description,
+            n_samples=len(shard_descriptor),
+            sample_shape=shard_descriptor.sample_shape,
+            target_shape=shard_descriptor.target_shape
+        )
+
+        shard_info.node_info.CopyFrom(self._get_node_info())
+
+        acknowledgement = self.stub.AcknowledgeShard(shard_info)
+        return acknowledgement.accepted
+
+    def get_experiment_data(self):
+        """Get an experiment data from the director."""
+        logger.info('Send WaitExperiment request')
+        response_iter = self.stub.WaitExperiment(self._get_experiment_data())
+        logger.info('WaitExperiment response has received')
+        experiment_name = None
+        for response in response_iter:
+            experiment_name = response.experiment_name
```

Comment:
Here we want to get the last experiment name from iterator. If we want to get only one 
```python 
response = next(response_iter)
```
 can be used.